streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java

```diff
@@ -0,0 +1,172 @@
+ * The keys of the {@code Map<String, KStream<K, V>>} entries returned by {@link BranchedKStream#defaultBranch(Branched)} or
+ * {@link BranchedKStream#noDefaultBranch()} are defined by the following rules:
+ * <p>
+ * <ul>
+ *     <li>If {@link Named} parameter was provided for {@link KStream#split(Named)}, its value is used as
+ *     a prefix for each key. By default, no prefix is used
+ *     <li>If a name is provided for the {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, its value is appended to the prefix to form the {@code Map} key
+ *     <li>If a name is not provided for the branch, then the key defaults to {@code prefix + position} of the branch
+ *     as a decimal number, starting from {@code "1"}
+ *     <li>If a name is not provided for the {@link BranchedKStream#defaultBranch()} call, then the key defaults
+ *     to {@code prefix + "0"}
+ * </ul>
+ * <p>
+ * The values of the respective {@code Map<Stream, KStream<K, V>>} entries are formed as following:
+ * <p>
+ * <ul>
+ *     <li>If no chain function or consumer is provided {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, then the value is the branch itself (which is equivalent to {@code ks -> ks}
+ *     identity chain function)
+ *     <li>If a chain function is provided and returns a non-null value for a given branch, then the value is
+ *     the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then the respective entry is not put to the map.
+ *     <li>If a consumer is provided for a given branch, then the the respective entry is not put to the map
+ * </ul>
+ * <p>
+ * For example:
+ * <pre> {@code
+ * Map<String, KStream<..., ...>> result =
+ *   source.split(Named.as("foo-"))
+ *     .branch(predicate1, Branched.as("bar"))                    // "foo-bar"
+ *     .branch(predicate2, Branched.withConsumer(ks->ks.to("A"))  // no entry: a Consumer is provided
+ *     .branch(predicate3, Branched.withFunction(ks->null))       // no entry: chain function returns null
+ *     .branch(predicate4)                                        // "foo-4": name defaults to the branch position
+ *     .defaultBranch()                                           // "foo-0": "0" is the default name for the default branch
+ * }</pre>
+ *
+ * <h2><a name="examples">Usage examples</a></h2>
+ *
+ * <h3>Direct Branch Consuming</h3>
+ * In many cases we do not need to have a single scope for all the branches, each branch being processed completely
+ * independently from others. Then we can use 'consuming' lambdas or method references in {@link Branched} parameter:
+ *
+ * <pre> {@code
+ * source.split()
+ *     .branch((key, value) -> value.contains("A"), Branched.withConsumer(ks -> ks.to("A")))
+ *     .branch((key, value) -> value.contains("B"), Branched.withConsumer(ks -> ks.to("B")))

```

Comment:
To simplify the example, should we use `predicate1`, `predicate2` as in the first example?